### PR TITLE
[WIP] Reservoir sampling (unweighted + weighted)

### PR DESCRIPTION
**Pull request by atavory**
Data structures for reservoir sampling (unweighted + weighted).

The implementation aims to allow generic sampling across all types, but to performa as small as possible a number of boxing and unboxing. It utilizes the fact that as streams are processed, elements selected for samples become increasingly rare.

### CODE REVIEW

**Comment by atavory** [comment]
@mbasmanova I'd appreciate some help on writing an array output. In `ReservoirSamplingAggregationFunction.java`, 238, the call to `BlockBuilder.beginBlockEntry` fails, with an `UnsupportedOperationException`. Not sure why that is the case, nor what is the correct way to "open an array", followed by writing things into it in a loop, followed by "closing it".

Many thanks!

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/ReservoirSamplingAggregationFunction.java [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+                generateAggregationName(getSignature().getName(), type.getTypeSignature(), inputTypes.stream().map(Type::getTypeSignature).collect(toImmutableList())),
+                parametersMetadata,
+                inputFunction,
+                combineFunction,
+                outputFunction,
+                ImmutableList.of(new AccumulatorStateDescriptor(
+                        stateInterface,
+                        stateSerializer,
+                        stateFactory)),
+                type);
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/ReservoirSamplingAggregationFunction.java]: @atavory 

> @mbasmanova I'd appreciate some help on writing an array output. In ReservoirSamplingAggregationFunction.java, 238, the call to BlockBuilder.beginBlockEntry fails, with an UnsupportedOperationException. Not sure why that is the case, nor what is the correct way to "open an array", followed by writing things into it in a loop, followed by "closing it".
> 
> Many thanks!

The problem is that the aggregation's declared return type is wrong. It is `type`, but needs to be `new Array(type)` on L172.

Comment (atavory) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/ReservoirSamplingAggregationFunction.java]: Many thanks! Just to be sure, you meant `new ArrayType(type)`, right?

Comment (mbasmanova) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/ReservoirSamplingAggregationFunction.java]: @atavory 

> Just to be sure, you meant new ArrayType(type), right?

That's right.

**Comment by atavory** [comment]
@mbasmanova 

Thanks for your help regarding this.

After some thought, I think that, in terms of order, the data structures for reservoir sampling are more useful for building blocks for other UDFS (such as differential entropy), than as standalone UDFS. Consequently, I've split the diff so that this part contains only the data structures, and later diffs will use them (including the reservoir sampling UDFs themselves).

I still need to add (many) more unittests, but would appreciate it if you could review. Many thanks!

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/UnweightedReservoirSample.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+package com.facebook.presto.operator.aggregation.reservoirsample;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.concurrent.ThreadLocalRandom;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+public class UnweightedReservoirSample
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/UnweightedReservoirSample.java]: @atavory Ami, I'd like to see at least one usage of these data structures. They store values in `List<Object>` regardless of the actual data type. This is a lot more expensive than necessary if underlying data type is integer, long or boolean. For these data types, using primitive arrays (int[], long[], boolean[]) is way more efficient and GC-friendly. How many objects do you expect to store? Do we need to copy these list across machines?

Comment (atavory) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/UnweightedReservoirSample.java]: @mbasmanova 

Thanks for comment.

What is the alternative? I could duplicate the code for each primitive type and use primitive arrays, but, AFAICT about Java (in which I'm not an expert), there is no way to do this generically otherwise.

As to your other questions: IDK how many objects will be asked to be stored (it will be specified by people calling the UDF, much like `numeric_histogram`). The lists will be transferred across machines (I haven't yet written the serialization code).

Comment (mbasmanova) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/UnweightedReservoirSample.java]: @atavory 

> What is the alternative? I could duplicate the code for each primitive type and use primitive arrays

It may be necessary to write separate implementations for each type.

> IDK how many objects will be asked to be stored (it will be specified by people calling the UDF, much like numeric_histogram). The lists will be transferred across machines (I haven't yet written the serialization code).

This is important because we have a limit of 16MB for a chunk of data being transferred. Hence, compact serialization might be important.

Let's start with a specific use case (a UDF) and figure all this out.

Comment (atavory) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/UnweightedReservoirSample.java]: @mbasmanova Still adding unittests, but addressed your comment above, I hope. Thanks!

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/WeightedDoubleReservoirSample.java [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    {
+        return new WeightedDoubleReservoirSample(this);
+    }
+
+    public DoubleStream stream()
+    {
+        return Arrays.stream(elements);
+    }
+
+    private void assertValid()
```
Comment (atavory) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/WeightedDoubleReservoirSample.java]: @mbasmanova 

What's the Presto way to mark code as debug-mode only?

Comment (mbasmanova) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/WeightedDoubleReservoirSample.java]: @atavory There is no such way. Why do you need this?

**Comment by mbasmanova** [comment]
@atavory 

> Still adding unittests, but addressed your comment above, I hope. 

I don't see any use case included in the PR. Having a specific use case is essential in deciding on the design of the helper classes. Am I missing something?

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/WeightedDoubleReservoirSample.java (hunk 3 above)
Comment (atavory) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/WeightedDoubleReservoirSample.java]: To check the heap invariants. The check is a linear-time operation (which makes the heap meaningless). 

I will remove this from the code.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/UnweightedDoubleReservoirSample.java [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+    }
+
+    public long getMaxSamples()
+    {
+        return maxSamples;
+    }
+
+    public void add(double element)
+    {
+        int index = getInsertIndex();
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/UnweightedDoubleReservoirSample.java]: inline getInsertIndex and putAtIndex

```
        ++count;

        int index;
        if (elements.length < maxSamples) {
            index = elements.length;
        }
        else { 
            index = ThreadLocalRandom.current().nextInt(0, (int) count);
        }
        
        if (index < elements.length) {
            if (elements.length < maxSamples) {
                double[] newElements = new double[elements.length + 1];
                System.arraycopy(elements, 0, newElements, 0, elements.length);
                elements = newElements;
            }

            elements[index] = element;
        }
```

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/UnweightedDoubleReservoirSample.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+            System.arraycopy(elements, 0, newElements, 0, elements.length);
+            newElements[elements.length] = element;
+            elements = newElements;
+            return;
+        }
+
+        elements[index] = element;
+    }
+
+    public void mergeWith(UnweightedDoubleReservoirSample other)
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/UnweightedDoubleReservoirSample.java]: Why is this needed?

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/UnweightedDoubleReservoirSample.java [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+        elements = merged;
+    }
+
+    @Override
+    public UnweightedDoubleReservoirSample clone()
+    {
+        return new UnweightedDoubleReservoirSample(this);
+    }
+
+    public DoubleStream stream()
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/UnweightedDoubleReservoirSample.java]: Why is this needed?

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/UnweightedDoubleReservoirSample.java [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+    }
+
+    protected void putAtIndex(int index, double element)
+    {
+        if (index == -1) {
+            return;
+        }
+
+        if (elements.length < maxSamples) {
+            double[] newElements = new double[elements.length + 1];
```
Comment (mbasmanova) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/UnweightedDoubleReservoirSample.java]: - Re-allocating the array and copying data for each element is wasteful; why not allocate maxSamples in the constructor?

```
    public UnweightedDoubleReservoirSample(int maxSamples)
    {
        checkArgument(maxSamples > 0, "Number of reservoir samples must be strictly positive");
        checkArgument(maxSamples < MAX_SAMPLES_LIMIT, "Number of reservoir samples must not exceed " + MAX_SAMPLES_LIMIT);
        this.maxSamples = maxSamples;
        this.elements = new double[maxSamples];
    }
```

**Comment by atavory** [comment]
@mbasmanova 

OK, thanks. I was hoping to split the diff into one involving only the reservoir-sample data structures, and one involving a continuous-sample entropy UDF, but sure. I'll mark this as WIP for now, and will ask for a review later on, when I rebase this on top of other diffs. Thanks!

**Comment by mbasmanova** [comment]
@atavory FYI, you can have a single PR with multiple commits.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/UnweightedDoubleReservoirSample.java [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+        implements Cloneable
+{
+    protected int maxSamples;
+    protected long count;
+    double[] elements;
+
+    public UnweightedDoubleReservoirSample(int maxSamples)
+    {
+        this.maxSamples = maxSamples;
+        checkArguments();
```
[the remark this answers is not in the document]
Comment (atavory) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/UnweightedDoubleReservoirSample.java]: I don't quite understand this comment. This constant is in some obscure inner class in the source code of general Presto. How do we know how much memory is running on other installations of Presto? At our installations in 5 years? 

The type of constant you're suggesting doesn't seem to me to belong in the code, but rather by some tool used by someone installing and maintaining Presto. They can use cgroups for that without modifying the source code.

[the remark this answers is not in the document]
Comment (mbasmanova) [reply presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/UnweightedDoubleReservoirSample.java]: @atavory 

> They can use cgroups for that without modifying the source code.

Presto is a multi-tenant system, e.g. a single JVM is processing multiple queries at the same time. Cgroup configuration can limit the total amount of memory used by all the queries on a single host, but can't be used to limit amount of memory consumed by individual queries. Hence, it is important to ensure that no single query can use all of the memory on a host starving other queries and causing instability. This is the main reason why Presto cannot support UDFs that use external resources (make network calls, read from disk, allocate large chunks of memory).

It might be reasonable to make `MAX_SAMPLES_LIMIT` value configurable, however, it would be very hard for the system administrator to find the right value. Allowing any query to use way more memory than others increases the risk of having reliability problems.

There must be a limit. The default value for that limit should be such that it makes sense in most cases. If it is hard to come up with default value that works in 80% of use cases, then, a configuration option to change that limit might be helpful.

For the use cases you know about, what's the number of samples they require?

**Comment by atavory** [comment]
Per @mbasmanova's comments, folded this into 13035
